joomla/joomla-cms
# [5.0][RFC][Events] Make captcha plugins use plugin events

Pull Request for Issue #28800 .

### Summary of Changes

The patch improve Captcha plugins.
Now it can use a **real events,** and people stop confusing `onDisplay` (etc) callback with event.

**How this work.**
Here added a new event `onCaptchaSetup` and an Interface for `CaptchaProvider` class.
While this event the plugin should register own `CaptchaProvider` class in `CaptchaRegistry`.
The system will retrieve the requested Captcha type from that registry and call required methods, that defined by `CaptchaProviderInterface` 

This is fully backward compatible.

Looking for feedback. 
If all good we can do the same for Editor and its Buttons.

*Techicaly now one plugin can provide multiple captcha, however for now the global configuration done per plugin.
But this can be extended in future, when need.*

### Testing Instructions

Apply patch, remove `administrator/cache/autoload_psr4.php`
Enable `recaptcha_invisible` and make sure it works.

Testing B/C:
Enable older `recaptcha` and make sure it works.

### Actual result BEFORE applying this Pull Request
All works

### Expected result AFTER applying this Pull Request
All works


### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: https://docs.joomla.org/Plugin/Events
- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/96
